### DISCUSSION TITLE
Character Export Schema for Optolith 2.x.x

### DISCUSSION THREAD

**Discussion by elyukai** [elyukai/optolith-character-schema#1] [Ideas]
This is a proposal for an updated character JSON export schema for Optolith version 2.x.x.

Feedback of any kind is welcome, let it be structure, naming, conventions, …! ❤

**Comment by JoveToo** [comment]
Since you are now redoing it all, would it be possible to add a log feature to the character? that shows what changes are made ot the character? (improving skill, add spells, ...)
it would make it very easy for me as a DM to keep track of what changes in the characters of my players

**Reply by soulflyman** [reply]
@JoveToo There is already a ticket for that feature elyukai/optolith-client#37, but you are right that would be an awesome feature to have. I think it can be implemented without depending on the structure of the character JSON

**Comment by elyukai** [comment]
> The "skills" currently do not contain the scope of application (“Anwendungsgebiete”). Adding these allows some improvements on the character sheet by adding more information there. I give one example in German language.
> 
> Lebensmittelbearbeitung – Backen, Braten&Sieden, Brauen, Haltbarmachen
> 
> There are some general special abilities (“Allgemeine Sonderfertigkeiten”) that extend the scope of application of a talent. Such extension could be made visible in the list of skills making them much easier to find and apply. Example in German language.
> 
> Winzerei – erweitert Lebensmittelbearbeitung
> Lebensmittelbearbeitung – Backen, Braten&Sieden, Brauen, Haltbarmachen, Winzerei
> 
> In my opinion having the scope extension visible on the skill itself is an improvement on the charactersheet where I would show such special abilities much less prominent.
> 
> In may also be possibile to store the scope of application somewhere else as it is more a part of the rules than a part of the character.

— [**mwberlindeu**](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3700465#gistcomment-3700465)

**Reply by elyukai** [reply]
> The "skills" currently do not contain specialisations (“Fertigkeitsspezialisierung”). Adding these allows some improvements on the character sheet by adding more information there, i.e. to set specialisations in bold. I give one example in German language.
>
> Fertigkeitsspezialisierung (Körperbeherrschung: Kampfmanöver)
> Körperbeherrschung - Akrobatik, Balance, Entwinden, Kampfmanöver, Laufen, Springen
>
> Strictly speaking though you do not necessarily have to move the information to the skills sections as the information is already contained in the special abilities (“Allgemeine Sonderfertigkeiten”).
> In my opinion having the skill specialisations visible on the skill itself is an improvement on the charactersheet where I would show skill specialisations under the special abilities much less prominent.

— [**mwberlindeu**](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3700474#gistcomment-3700474)

**Reply by elyukai** [reply]
> Thank you for your feedback! But … I'm a bit confused: You mention the character sheet, but this is not the character sheet …? 🙈

— [me](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3700604#gistcomment-3700604)

**Reply by elyukai** [reply]
> I know. I hinted at the end use with mentioning the charactersheet, and thought that you may have to do something on the Character JSON Schema to make the suggestions possible.

— [**mwberlindeu**](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3700616#gistcomment-3700616)

**Reply by elyukai** [reply]
> Ah, okay, well, the character JSON schema is only for the JSON export of Optolith to share characters with others or "third-party" tools. It does not have any influence on the character sheet or other UI, which is handled internally.

— [me](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3700772#gistcomment-3700772)

**Comment by elyukai** [comment]
> Maybe I missed it, but is it planed to have derived values for attack and parry of each weapon in the exported json?

— [**soulflyman**](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3733965#gistcomment-3733965)

**Reply by elyukai** [reply]
> > Maybe I missed it, but is it planed to have derived values for attack and parry of each weapon in the exported json?
>
> You didn't miss it, since items are not included there yet 😅🙈 – the database schema still needs to be finalized and it doesn't make much sense to work on this before.
>
> And no, I haven't considered including AT/PA/RC yet, although it would make sense! 😮 Thank you! I wonder, tho, if it would make more sense to wait with this for Item Sets, since then the value will be completely calculated. Currently it is not that clear which entries influence the values, but especially for tooling authors it would be important to know if a value is a "final" value, and if not, which exact influence entries have. At least this is what I guess.
>
> And I don't think it would make sense to include both versions of those values – with and without item set. And even tho I try to provide easier-to-use value calculations I also think it would make sense to not "overload" the documents with very fine-grained and detailed values. And, again, this is what I assume.
>
> What do *you* think about that?

— [me](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3733992#gistcomment-3733992)

**Reply by elyukai** [reply]
> > And I don't think it would make sense to include _both_ versions of those values – with and without item set. And even tho I try to provide easier-to-use value calculations I also think it would make sense to not "overload" the documents with very fine-grained and detailed values. And, again, this is what I assume.
> > 
> > What do _you_ think about that?
>
> Overloading or cluttering the export is not a good idea. I think of the derived values as a machine readable version of the printed character sheet. Everything that is calculated and printed on the character sheet should be in the export as derived value, everything else, that a player needs to take care off during game play, should not be printed/exported as calculated/derived value.
>
> So the derived AT/PA value should match that values that are printed on the character sheet in the row of the weapon.
>
> For example Optodice, the dice rolling tool I develop, could just use the derived AT value that Optolith does export. If anything influences this I would expect that Optolith does take care of it and gives me a final value that I can work with. But I can just speak for my tool and I don't know if other tools have to know if the value is influenced by something else :shrug: 

— [**soulflyman**](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3734043#gistcomment-3734043)

**Reply by elyukai** [reply]
> Thanks for your input (again)! 
>
> Following that I would tend to not include AT/PA values for separate weapons and instead wait for items sets so that you'll definitely have final values.

— [me](https://gist.github.com/elyukai/9152a4985c2c1384840eeadaf451d848?permalink_comment_id=3735291#gistcomment-3735291)
